walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,206 @@
+
+    // If no mandatory attendees are passed, return the available TimeRanges for optional attendees
+    // else if no optional attendees are passed, or no TimeRanges are available for optional
+    // attendees then return available TimeRanges for mandatory attendees, else merge the available
+    // TimeRanges for mandatory attendees with available TimeRanges for optional employees.
+    if (mandatoryAttendees.isEmpty()) {
+      return availableOptionalTimeRanges;
+    } else if (optionalAttendees.isEmpty() || availableOptionalTimeRanges.isEmpty()) {
+      return availableMandatoryTimeRanges;
+    } else {
+      return mergeAvailableTimeRanges(
+          availableMandatoryTimeRanges, availableOptionalTimeRanges, meetingDurationMinutes);
+    }
+  }
+
+  /** Returns a List of TimeRanges that are unavailable in as little TimeRanges as possible. */
+  private List<TimeRange> getUnavailableTimeRanges(
+      Collection<Event> events, Collection<String> attendees) {
+    List<TimeRange> unavailableTimeRanges = new ArrayList<>();
+
+    // If an event's attendee list contains a names of an attendee that needs to attend
+    // (from Collection<String> attendees, then we add the event's TimeRange to a list.
+    events.forEach(
+        event -> {
+          Collection<String> eventAttendees = event.getAttendees();
+
+          for (String attendee : eventAttendees) {
+            if (!attendees.contains(attendee)) {
+              continue;
+            }
+
+            TimeRange currentEventTimeRange = event.getWhen();
+
+            // If the currentEventTimeRange isn't merged into a TimeRange within the
+            // unavailableTimeRanges list, then add the currentEventTimeRange to the list.
+            if (!mergeOnOverlap(unavailableTimeRanges, currentEventTimeRange)) {
+              unavailableTimeRanges.add(currentEventTimeRange);
+            }
+
+            break;
+          }
+        });
+
+    return unavailableTimeRanges;
+  }
+
+  /**
+   * Returns whether a merge has ocurred or not and modifies the Collection of TimeRanges to include
+   * a merged TimeRange where the newTimeRange overlaps with the passed Collection.
+   */
+  private boolean mergeOnOverlap(Collection<TimeRange> timeRanges, TimeRange newTimeRange) {
+    // The default value is set to false, which is converted to true if a merge takes place.
+    boolean doOverlap = false;
+
+    // Store items that we will add and remove from the collection after we finish iterating
+    // through it in order to avoid ConcurrentModificationException from being thrown.
+    List<TimeRange> toRemove = new ArrayList<>();
+    List<TimeRange> toAdd = new ArrayList<>();
+
+    for (TimeRange takenTimeRange : timeRanges) {
+      if (newTimeRange.overlaps(takenTimeRange)) {
+        TimeRange combinedTimeRange = mergeTimeRanges(newTimeRange, takenTimeRange);
+        toRemove.add(takenTimeRange);
+        toAdd.add(combinedTimeRange);
+        doOverlap = true;
+      }
+    }
```

Comment:
I think with this logic and this example:
```
existingRanges: ----|******|-------|******|---
newTimeRange:   --------|************|--------
```
you'll end up with two time ranges that overlap in the result. Is that okay, because it seems like it's what you're trying to avoid by doing this merge in the first place?

Reply:
Done.